oe_contact_forms.module

```diff
@@ -355,3 +359,100 @@ function _oe_contact_forms_email_validate(array &$form, FormStateInterface $form
+  /** @var \Drupal\contact\MessageInterface $contact_message */
+  $contact_message = $form_object->getEntity();
+  /** @var \Drupal\contact\Entity\ContactForm $contact_form */
+  $contact_form = ContactForm::load($contact_message->bundle());
+
+  if (!$contact_form->getThirdPartySetting('oe_contact_forms', 'is_corporate_form', FALSE)) {
+    $fields = ['oe_country_residence', 'oe_telephone', 'oe_topic'];
+
+    foreach ($fields as $field) {
+      $form[$field]['#access'] = FALSE;
+    }
+  }
+}
+
+/**
+ * Implements hook_entity_base_field_info().
+ */
+function oe_contact_forms_entity_base_field_info(EntityTypeInterface $entity_type) {
+  if ($entity_type->id() == 'contact_message') {
+    $fields = [];
+
+    $fields['oe_country_residence'] = BaseFieldDefinition::create('skos_concept_entity_reference')
+      ->setLabel(t('Country of residence'))
+      ->setSetting('target_type', 'skos_concept')
+      ->setSetting('handler', 'default:skos_concept')
+      ->setSetting('handler_settings', [
+        'target_bundles' => NULL,
+        'auto_create' => FALSE,
+        'concept_schemes' => [
+          'http://publications.europa.eu/resource/authority/country',
+        ],
+      ])
+      ->setDisplayOptions('form', [
+        'type' => 'skos_concept_entity_reference_autocomplete',
+      ])
+      ->setDisplayOptions('view', [
+        'weight' => 0,
+      ]);
+
+    $fields['oe_telephone'] = BaseFieldDefinition::create('telephone')
+      ->setLabel(t('Phone'))
+      ->setSettings([
+        'default_value' => '',
+        'max_length' => 255,
+      ])
+      ->setDisplayOptions('form', [
+        'type' => 'tel',
+      ])
+      ->setDisplayOptions('view', [
+        'weight' => 0,
+      ]);
+
+    $fields['oe_topic'] = BaseFieldDefinition::create('list_string')
+      ->setLabel(t('Topic'))
+      ->setDisplayOptions('form', [
+        'type' => 'options_select',
+      ])
+      ->setDisplayOptions('view', [
+        'weight' => 0,
+      ])
+      ->setSetting('allowed_values_function', '_oe_contact_forms_topic_values')
+      ->setRequired(TRUE);
+
+    return $fields;
+  }
+}
+
+/**
+ * Set dynamic allowed values for the topic field.
+ *
+ * @param \Drupal\Core\Field\BaseFieldDefinition $definition
+ *   The base field definition.
+ * @param \Drupal\contact\Entity\Message|null $entity
+ *   The entity being created if applicable.
+ * @param bool $cacheable
+ *   Boolean indicating if the results are cacheable.
+ *
+ * @return array
+ *   An array of possible key and value options.
+ *
+ * @see options_allowed_values()
+ */
+function _oe_contact_forms_topic_values(BaseFieldDefinition $definition, Message $entity = NULL, &$cacheable = TRUE): array {
+  /** @var \Drupal\contact\Entity\ContactForm $contact_form */
+  $contact_form = $entity->getContactForm();
+  $topics = $contact_form->getThirdPartySetting('oe_contact_forms', 'topics', []);
+  $options = array_column($topics, 'topic_name');
```

Comment:
No keys? We key them numerically here? Doubt this is good.

Reply:
they come in sets of name and mail, with the very same keys, so there will not be any issues

Reply:
No, this is not correct! Array_column will turn them into numeric keys so in the DB for the topic you are storing 0 or 1 or 2 for the topic! Can even see it in the test. This is not good:

![Screenshot 2020-05-14 08 50 00](https://user-images.githubusercontent.com/5848933/81902147-f0d4ee00-95bf-11ea-85e4-27aa353ed895.png)